lib/src/kbasesearchengine/events/AccessGroupEventQueue.java

```diff
@@ -0,0 +1,314 @@
+ * 
+ * This class is not thread safe.
+ * @author gaprice@lbl.gov
+ *
+ */
+public class AccessGroupEventQueue {
+    
+    private static final Set<StatusEventType> ACCESS_GROUP_EVENTS = new HashSet<>(Arrays.asList(
+            StatusEventType.COPY_ACCESS_GROUP, StatusEventType.DELETE_ACCESS_GROUP,
+            StatusEventType.PUBLISH_ACCESS_GROUP)); 
+
+    private final Map<String, ObjectEventQueue> objectQueues = new HashMap<>();
+    
+    private final PriorityQueue<StoredStatusEvent> accessGroupQueue =
+            new PriorityQueue<StoredStatusEvent>(new Comparator<StoredStatusEvent>() {
+                
+                @Override
+                public int compare(final StoredStatusEvent e1, final StoredStatusEvent e2) {
+                    return e1.getEvent().getTimestamp().compareTo(e2.getEvent().getTimestamp());
+                }
+            });
+    /* if drain != null, ready and processing must = null. Object queues should be draining
+     * events until the timestamp on the drain event.
+     * if ready or processing != null, drain and the non null field must be null. All object
+     * queues must be drained at this point. 
+     */
+    private StoredStatusEvent drain = null; // AG event waiting for object queues to drain
+    private StoredStatusEvent ready = null; // AG event ready for processing
+    private StoredStatusEvent processing = null; // AG event processing
+    private int size = 0; // keep a record of size rather than running through sub queues
+    
+    /* should maybe initialize with an acccess group id and reject events that don't match */
+    
+    /** Create a new, empty, queue. */
+    public AccessGroupEventQueue() {}
+    
+    /** Create a new queue and initialize it with events in the
+     * {@link StatusEventProcessingState.READY} and {@link StatusEventProcessingState.PROC} states.
+     * 
+     * The set of events cannot violate the rules that
+     * <ul>
+     * <li> Only one access group level event can be in the ready or processing state, and
+     * if so no other events can be in the ready or processing state</li>
+     * <li> Only one object level event per object ID can be in the ready or processing state, and
+     * if so no version level events for that objectID can be in the ready or processing state</li>
+     * </ul>
+     * @param initialLoad the events to load into the queue.
+     */
+    public AccessGroupEventQueue(final Collection<StoredStatusEvent> initialLoad) {
+        Utils.nonNull(initialLoad, "initialLoad");
+        Utils.noNulls(initialLoad, "initialLoad has null entries");
+        final Map<String, StoredStatusEvent> objects = new HashMap<>();
+        final Map<String, List<StoredStatusEvent>> objectVersionReady = new HashMap<>();
+        final Map<String, List<StoredStatusEvent>> objectVersionProcessing = new HashMap<>();
+        
+        for (final StoredStatusEvent e: initialLoad) {
```

Comment:
coding style: In general I don't like using single letter variables like 'e' because it becomes hard to search through code for the variable. I tend to use acronyms 'sse' or just 'ee'. I don't expect you to go through and change all single letter vars as there are way too many to change, but just mentioning it as a good practice that can be followed in the future.

Reply:
My rules for single letter vars are that they need to be isolated to a < 10 line block or so, so it's trivial to see the entire scope of the variable. I think in that case it's ok.